### ISSUE TITLE
[FEATURE] A way to create custom constructors (where the field(s) to include are explicitly indicated)

### ISSUE THREAD

**Issue by rzwitserloot** [projectlombok/lombok#2957]
## Background and description

Lombok currently offers 3 'flavours' of constructor generation, all of which pre-define which fields are intended (in that they are named `NoArgsConstructor`, `AllArgsConstructor`, and `RequiredArgsConstructor`, which are words that describe which fields are to be included inherently). One of these features is wound up with `@Builder` (which implies `@AllArgsConstructor` under most scenarios).

There are multiple feature requests for more fine grained control of which fields are to be included, generally in the vein of 'how can I mark a field that lombok thinks is required as not actually required, or vice versa', or 'how can I exclude a field from `@Builder`?'.

These feature requests do not make sense, it's like asking 'how do I add a corner to a circle'. Nevertheless, the underlying requirement is a use case we'd like to cater to, so we need some sort of 4th flavour of constructor: Based on a custom list.

One annoyance is that java allows multiple constructors in a class, so it's not really right to talk about __the__ custom constructor; one can merely talk about __a__ custom constructor. This does not mesh well with annotations, and doesn't mesh well with the most obvious approach (which is to mark each field you want included explicitly, but how do you indicate that some field should be included for custom constructor 1, and not custom constructor 2?)

## Simple proposal - single `@CustomConstructor`

We ditch the use case of multiple custom constructors, relegating that to the 'that is no longer boilerplate; lombok cannot help you out anymore' dustbin. In this case, all we really need is [A] a `@CustomConstructor` annotation  (valid on types) and [B] a `@CustomConstructor.Include` and `@CustomConstructor.Exclude` annotation (valid on fields).

Obvious downside is that you can only have 1 custom constructor.

It is a compiler error if you have include/exclude annotations and no `@CustomConstructor` on the class. It is a compiler error if you have `@CustomConstructor` on a class and zero include/exclude annotations (because, does that mean 'I want no fields' or does that mean 'I want all of them').

## Denied alternative: Omit `@CustomConstructor` itself (only have Incl/Excl)

It seems like the `@CustomConstructor` annotation itself doesn't really need to exist; a single Include or Exclude annotation is enough. However, that complicates matters a bit for `@Builder` and co, and I don't think it helps readability of the end result, nor does it help discoverability of the `@CustomConstructor` feature. Unless I've missed something (in which case feel free to comment), I'm denying this alternative.

## Denied alternative: Stringly typed multiple custom constructors

We _could_ introduce the concept of the 'constructor name'. Then you could do something like:

```
@CustomConstructor({"a", "foo"})
public class Example {
    @CustomConstructor.Include("a")
    private int includedInAButNotInFoo;
}
```

but this seems really hard to use, easily fatfingered, lacks IDE support (auto-complete on those constructor names is not going to happen, obviously), and even causes confusion when you want to define one constructor by way of inclusion (only these fields), and the other by way of exclusion (all fields except the ones marked for exclusion).

Making this feature way more complicated in the name of what to me seems like a really exotic usecase (multiple constructors required, all of which are field-based and none of which fall under the notion of 'all', 'none', or 'required'), doesn't sound like a good plan. Unless there's a common pattern/very commonly used library out there that more or less requires at least 2 custom constructors that I'm not aware of (in which case, comment and name them), this isn't going to happen (especially not for a first take on this idea; we can always add it later without breaking backwards compatibility if we must).

### What about final fields?

What do we do with non-included `final` fields? Just init them to 0/false/null? Allow a mechanism to specify the default (this is incredibly complicated, so before you chime in: YEAH! It makes this feature about 10x more complicated, I'd like to hear a clear proposal on how. Any comments with a simple 'yes, important!', even if it goes with a personal story on how much you need it, are not particularly desired, I don't think anybody would argue defaults aren't useful). This defaulting question also extends to NoArgsConstructor to some extent.

* Just use a single default annotation doesn't work - java's annotation system doesn't allow e.g. `@Default(System.currentTimemillis())` at all (only compile time constants need apply). Even if we restrict defaults to only CTCs, you'd need one annotation for each type (`@DefaultInt`, `@DefaultBoolean`, etc).
* Allowing the field initializer to serve (the way builder works: `@Default private int field = 5;`, implying that only for those constructors that do not include `field`, set it to `5`) is also complicated: What if that expression has side-effects or resolves to a mutable entity. Does that mean it's resolved once for the entire class and that value is assigned as default, or is the expression resolved every time? We also have to 'mess' with your hand written code and move that initializing expression around, which makes this feature very complex and significantly increases maintenance burden. We'd also have to answer some questions about having 2 separate defaulting mechanisms - this really doesn't sound like a workable idea unless we unify `@Builder.Default` and this `@Default` first, but how do we do that, preferably maintaining backwards compatibility?

I will assume that this feature will be implemented separated (as in, we'll ship without defaulting support now; we can always add it later).

Status of this feature request: Gather feedback and use cases first, before making a decision.

**Comment by liampace** [comment]
> There are multiple feature requests for more fine grained control of which fields are to be included, generally in the vein of 'how can I mark a field that lombok thinks is required as not actually required, or vice versa', or 'how can I exclude a field from `@Builder`?'.
> 
> These feature requests do not make sense, it's like asking 'how do I add a corner to a circle'. 

I have to disagree with this take. It makes intuitive sense for transient fields to not be included as by their nature they are dependent on the runtime state of the program and as such should not be arguments for constructors. Take for instance a simplified `App` class:
```
public class App implements Runnable {
    @Singular private List<Runnable> tasks = new ArrayList<>();
    @Getter private transient boolean running = false;

    public void quit() {
        if (!this.running)
            throw new IllegalStateException("App was not running");
        this.running = false;
    }

    @Override
    public void run() {
        this.running = true;
        while (this.running) {
             // ...
        }
    }
}
```
In this class the `running` field is completely managed internally and the user should never be able to directly manipulate it. However, when using the `Builder` annotation the user is able to directly manipulate the state of this field. On more complex classes, giving the user direct access to transient fields can lead to corrupted state or worse breaking the defined API/ABI contracts.

**Comment by liampace** [comment]
@rzwitserloot 
> ### What about final fields?
> 
> What do we do with non-included `final` fields? Just init them to 0/false/null?

Here are two potential solutions off the top of my head:

# 1. Inclusion/Exclusion for Non-Final Fields Only
Instead of having the `@*.Include`/`@*.Exclude` applying to every field, what if we make it only apply to non-final fields? This way we don't have to worry about setting default values for the final fields. By doing this, the `@CustomConstructor` would act as an intermediate step between `@RequiredArgsConstructor` and `@AllArgsConstructor`. 

For example:
```java
@AllArgsConstructor
@CustomArgsConstructor
class Example {
    private final String a;
    @CustomArgsConstructor.Include private long b;
    @CustomArgsConstructor.Exclude private int c;
}
```
would result in:
```java
class Example {
    private final String a;
    private long b;
    private int c;

    public Example(final String a, final long b, final int c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }

    public Example(final String a, final long b) {
        this.a = a;
        this.b = b;
    }
}
```

Although if we were to go in this direction and depending on how we implement it (inclusion by default or exclusion by default) either `@*.Include` or `@*.Exclude` would become redundant. To get around this new problem we could have the `@CustomArgsConstructor` produce a warning if there are missing `@*.Include`/`@*.Exclude` on non-final fields.

---

# 2. `@CustomArgsConstructor` `force` Parameter (a la `@NoArgsConstructor`)
Just like `@NoArgsConstructor`, we can introduce a `force` parameter that when enabled automatically sets excluded final fields to `0`/`false`/`null`.  By doing it this way, we allow for more customization as it words on final fields, and it falls more in line with how lombok already deals with similar scenarios.

For example:
```java
@CustomArgsConstructor(force = true)
class Example {
    private final String a;
    @CustomArgsConstructor.Exclude private final long b;
}
```
would result in 
```java
class Example {
    private final String a;
    private final long b;

    public Example(final String a) {
        this.a = a;
        this.b = 0;
    }
}
```
The one caveat for this is that when `force` is set to false. In this scenario, I think we should produce an error stating that exclusion of final fields are only valid when the force parameter is set.

**Comment by liampace** [comment]
Another question to contend with is if we want to follow in the footsteps of `@ToString` and `@EqualsAndHashCode` by having an `onlyExplicitlyIncluded` parameter or not.

**Comment by Akhil-1005** [comment]
Hi, I'd like to implement this feature!
After reading through this issue and the maintainer's analysis, I want to move forward with an implementation following the design already outlined here.
Planned Design (based on @rzwitserloot's suggestion):
Class-level annotation + field-level include:
java@CustomConstructor
public class User {
    @CustomConstructor.Include
    private String name;
    @CustomConstructor.Include
    private String email;
    private int age; // excluded from constructor
}
Which would generate:
javapublic User(String name, String email) {
    this.name = name;
    this.email = email;
}
Edge cases I plan to handle:

Compiler error if @CustomConstructor is on a class with zero @Include annotations
Compiler error if @CustomConstructor.Include is used on a field without @CustomConstructor on the class
Non-included final fields will raise a compile error (must be initialized inline)

Before I open a PR, I just want to confirm:

Is this design still the preferred approach?
Should this support repeatable annotations (for multiple custom constructors on the same class)?

Happy to adjust based on feedback. Thanks
